rfcs/45-stageless.md

```diff
@@ -0,0 +1,977 @@
+- **dependency**: system (set) that must complete before another system (set) can run
+- **schedule** (noun): the executable form of a system set
+- **schedule** (verb): specify when and under what conditions systems run
+- **scheduler**: the programmer, who is attempting to specify when and under what conditions systems run
+- **executor**: runs the systems in a schedule on a world
+- **"ready"**: when a system is no longer waiting for dependencies to complete
+
+To write a Bevy app, you have to specify when your systems run.
+By default, systems have neither strict execution order nor any conditions for execution.
+**Scheduling** is the process of supplying those properties.
+To make things more ergonomic, systems can be grouped under **system sets**, which can be ordered and conditioned in the same manner as individual systems.
+This allows you to easily refer to many systems and (indirectly) give properties to many systems.
+Furthermore, systems and system sets can be ordered together and even grouped together *within larger sets*, meaning you can layer those properties.
+
+In short, for any system or system set, you can:
+
+- define its execution order relative to other systems or sets (e.g. "this system runs before A")
+- define any conditions that must be true for it to run (e.g. "this system only runs if a player has full health")
+- define which set(s) it belongs to (which define properties that affect all systems underneath)
+  - if left unspecified, the system or set will be added under a default one (configureable)
+
+These properties are all additive.
+Adding another does not replace an existing one, and they cannot be removed.
+
+### Sample
+
+```rust
+use bevy::prelude::*;
+
+#[derive(State)]
+enum GameState {
+    Running,
+    Paused
+}
+
+#[derive(SystemLabel)]
+enum MySystemSets {
+    Update,
+    Menu,
+    SubMenu,
+}
+
+#[derive(SystemLabel)]
+enum MySystems {
+    FlushMenu,
+}
+
+fn main() {
+    App::new()
+        /* ... */
+        .add_set(
+            // Use the same builder API for scheduling systems and system sets.
+            MySystemSets::Menu
+                // Put sets in other sets.
+                .in_set(MySystemSets::Update)
+                // Attach conditions to system sets.
+                // (If this fails, all systems in the set will be skipped.)
+                .run_if(state_equals(GameState::Paused))
+        )
+        .add_system(
+            some_system
+                .in_set(MySystemSets::Menu)
+                // Attach multiple conditions to this system, and they won't conflict with Menu's conditions.
+                // (All of these must return true or this system will be skipped.)
+                .run_if(some_condition_system)
+                .run_if(|value: Res<Value>| value > 9000)
+        )
+        // Bulk-add systems and system sets with some convenience macros.
+        .add_systems(
+          chain![
+            MySystemSets::SubMenu,

```

Comment:
Yeah, sorry the example isn't super clear. 

Basically, I wanted to show off that the macros can accept system sets too. It's not hard to convert every element into its descriptor and than wrap each descriptor in an enum so they're all the same type.

As you can write `my_system.after(my_system_set)` and vice-versa, it'd be weird if this convenience macro only worked for systems.

What you're suggesting could still be more idiomatic tho.

Reply:
The problem with that syntax is that I first thought it made `SubMenu` as a subset of `Menu`, until I tried to lay out a diagram of that. Then I realized that my mental model was wrong, because there was not an `add_set(MySystemSets::SubMenu)` anywhere.

Anyway, I don't really get the point of that, since `SubMenu` is just a label, so it doesn't have knowledge about what systems it is supposed to contain. That would force the user to actually add the system set with systems inside. Then adding the label like in the example above would re-add other instances of those systems. The more I think about it, the more it looks like a footgun to me.

Reply:
It sounds like there's another misunderstanding.

`add_systems` is equivalent to what you'd do if there was no shortcut. It wraps the descriptors like I said, then iterates them and calls `add_system` or `add_set` as needed.

You can add systems before or after adding their set. We can't check for errors until everything has been added, so the builder methods can be order-independent since graph construction can be deferred to the very end.

The only error we'd check for during the app builder phase is you adding two things with the same label. Since we refer to things by label for extraction, everything must have a unique label.

And nothing is ever re-added.

Reply:
Ok, understood, it just does an `add_set` internally. It's just the name that is confusing. Since it's called `add_systems`, one would expect it to just add systems.

Reply:
An earlier draft called the method `add_many`, but that was apparently too vague. I also think it supporting both systems and system sets wasn't clear to people then either tho, so idk.

There's no reason to limit it to systems.

We can discuss options later. It's possible to have a single `add` method that can accept both instead of `add_system` and `add_set` (but maybe call them `schedule` and `schedule_many` for clarity).